iocage/lib/JailState.py

```diff
@@ -0,0 +1,126 @@
+#    documentation and/or other materials provided with the distribution.
+#
+# THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+# IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+# IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+import typing
+import json
+import subprocess
+
+import iocage.lib.errors
+
+JailStatesDict = typing.Dict[str, 'JailState']
+
+
+def _parse_json(data: str) -> JailStatesDict:
+    output: typing.Dict[str, 'JailState'] = {}
+    jail_states = json.loads(data)["jail-information"]["jail"]
+    for jail_state_data in jail_states:
+        identifier = jail_state_data["name"]
+        output[identifier] = JailState(identifier, jail_state_data)
+    return output
+
+
+class JailState(dict):
+
+    name: str
+    _data: typing.Optional[typing.Dict[str, str]] = None
+    updated = False
+
+    def __init__(
+        self,
+        name: str,
+        data: typing.Optional[typing.Dict[str, str]]=None
+    ) -> None:
+
+        self.name = name
+
+        if data is not None:
+            self._data = data
+
+    def query(self) -> typing.Dict[str, str]:
+
+        data: typing.Dict[str, str] = {}
+        try:
+            stdout = subprocess.check_output([
+                "/usr/sbin/jls",
+                "-j",
+                self.name,
+                "-v",
+                "-h",
+                "--libxo=json"
+            ], shell=False, stderr=subprocess.DEVNULL)  # nosec TODO use helper
+            output = stdout.decode().strip()
+            data = _parse_json(output)[self.name]
+        except (subprocess.CalledProcessError, KeyError):
+            pass
+
+        self._data = data
+        return data
+
+    @property
+    def data(self) -> typing.Dict[str, str]:
+        if self._data is None:
+            self._data = self.query()
+        return self._data
+
+    def __getitem__(self, name: str) -> str:
+        return self.data[name]
+
+    def __iter__(
+        self
+    ) -> typing.Iterator[str]:
+        return self.data.__iter__()
+
+    def keys(self):
+        return self.data.keys()
+
+
+class JailStates(dict):
+
+    def __init__(
+        self,
+        states: typing.Optional[JailStatesDict]=None
+    ) -> None:
+
+        if states is None:
+            dict.__init__(self, {})
+        else:
+            dict.__init__(self, states)
+
+    def query(self):
```

Comment:
is this a standard `dict` function here?

Reply:
no, but update (my previous naming choice) was. This is called to query the state from jls.